docs/architecture/events.rst

```diff
@@ -44,3 +44,46 @@ Order events
 +--------------------------+--------------------------+---------------------------------------------------------------------+
 | ``other``                | ``OTHER``                | Status used during reimporting of the legacy events.                |
 +--------------------------+--------------------------+---------------------------------------------------------------------+
+
+Email Event Types
+-----------------
+
+Fill me...
+
+Events Design
+-------------
+
+The ``OrderEvent`` class contains a bunch of static methods suffixed by
+``_event``. They take care of all the logic and required fields. It whether
+returns an instance of ``OrderEvent``. This design has for objective to allow
+delayed (or not) event creation in the database.
+
+The main power of this design comes when one wants to create multiple
```

Comment:
This gives more flexibility. e.g. an order event generates a customer event. Dropping this would mean that simplicity may cost an additional query when writing to the DB but writes are not happening that often in e-commerce sites. Fair enough.

There is a case where we generate events from a for loop though. When we are bulk canceling orders. In that case, we could give it a special method, unless you see a better solution.

Reply:
I can imagine this might be a problem if the number of records is large and in fact, bulk actions are executed one by one. But I'm always for going with the simple approach first and optimizing when we identify the performance issue. I guess we'll need to soon do some performance tests to find out.